### PR TITLE
[FE-3064] Define an HTTPClient interface and default fetch client

### PR DESCRIPTION
**Pull request by ptpaterson**
Ticket(s): [FE-3064](https://faunadb.atlassian.net/browse/FE-3064)

This is an initial milestone to review a generic HTTPClient interface

[FE-3064]: https://faunadb.atlassian.net/browse/FE-3064?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/http-client/fetch-client.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          }
+        );
+      });
+
+    const status = response.status;
+
+    const responseHeaders: Record<string, string> = {};
+    response.headers.forEach((value, key) => (responseHeaders[key] = value));
+
+    const body: QuerySuccess<unknown> | QueryFailure = await response
```
Comment (cleve-fauna) [inline src/http-client/fetch-client.ts:43]:
```suggestion
    const body: QuerySuccess<JSONValue> | QueryFailure = await response
```

File: src/http-client/index.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  headers: Record<string, string>;
+  method: string;
+  url: string;
+  keepalive?: boolean;
+};
+
+export type HTTPResponse = {
+  headers: Record<string, string>;
+  status: number;
+  body: unknown;
```
Comment (cleve-fauna) [inline src/http-client/index.ts:14]: I suggest we add a new type - QueryResponse<T> = QuerySuccess<T> | QueryFailure; and then here do:

```suggestion
  body: QueryResponse<JSONValue>;
```

as fauna will only ever get us back JSON

**Review by cleve-fauna — COMMENTED** [review]
Suggestions on types and some interface bits.

File: src/http-client/index.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+export { FetchClient } from "./fetch-client";
+
+export type HTTPRequest = {
+  data: Record<string, any>;
+  headers: Record<string, string>;
+  method: string;
+  url: string;
+  keepalive?: boolean;
```
[the remark this answers is not in the document]
Comment (ptpaterson) [reply src/http-client/index.ts:8]: I'm good with punting on that. I'll cut another issue.

File: src/http-client/index.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+export { FetchClient } from "./fetch-client";
+
+export type HTTPRequest = {
+  data: Record<string, any>;
```
[the remark this answers is not in the document]
Comment (ptpaterson) [reply src/http-client/index.ts:4]: Hmmm. Actually, does it only accepts `QueryRequest`? I'll double check that and update.

[the remark this answers is not in the document]
Comment (cleve-fauna) [reply src/http-client/index.ts:4]: oh good call maybe so

File: src/http-client/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ptpaterson) [reply src/http-client/index.ts:13]: Ref [new Jira issue](https://faunadb.atlassian.net/browse/FE-3075)

File: src/http-client/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ptpaterson) [reply src/http-client/index.ts:9]: Updated to QueryRequest.

File: src/http-client/fetch-client.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    headers: requestHeaders,
+    method,
+    url,
+    keepalive,
+  }: HTTPRequest): Promise<HTTPResponse> {
+    // TODO: handle client timeouts with AbortController. Emit NetworkError if so.
+
+    const response = await fetch(url, {
+      method,
+      headers: requestHeaders,
```
[the remark this answers is not in the document]
Comment (ptpaterson) [reply src/http-client/fetch-client.ts:22]: done

**Review by cleve-fauna — APPROVED** [review]

**Pull request merged**